### PR TITLE
get rid of some invalidation sources in Artifacts

### PR DESCRIPTION
**Pull request by KristofferC**
Found with SnoopCompile and loading Plots.

Not sure if this is the best way to resolve them but it does the job. @timholy?

### CODE REVIEW

File: stdlib/Artifacts/src/Artifacts.jl [hunk 1]
```diff
@@ -270,7 +271,12 @@ function unpack_platform(entry::Dict, name::String, artifacts_toml::String)::Uni
     end
 
     # Collect all String-valued mappings in `entry` and use them as tags
-    tags = Dict(Symbol(k) => v for (k, v) in entry if isa(v, String))
```
Comment (timholy) [inline stdlib/Artifacts/src/Artifacts.jl:273]: `Symbol(k)::Symbol => v::String` might also work, but your version is good.

File: stdlib/Artifacts/src/Artifacts.jl [hunk 2]
```diff
@@ -610,24 +620,24 @@ end
 # Support `AbstractString`s, but avoid compilers needing to track backedges for callers
 # of these functions in case a user defines a new type that is `<: AbstractString`
 with_artifacts_directory(f::Function, artifacts_dir::AbstractString) =
-    with_artifacts_directory(f, string(artifacts_dir))
+    with_artifacts_directory(f, String(artifacts_dir)::String)
```
Comment (KristofferC) [inline stdlib/Artifacts/src/Artifacts.jl:623]: These are just fixes so that the code adheres to what the comment claims it does.

**Review by timholy — APPROVED** [review]

**Review by staticfloat — APPROVED** [review]
Thanks!

**Comment by staticfloat** [comment]
Looks like the keys to the dictionary passed to unpack_platform are not Strings, but possibly substrings.

**Comment by KristofferC** [comment]
> Looks like the keys to the dictionary passed to unpack_platform are not Strings, but possibly substrings.

I tweaked things a bit.

Basically, we have a "funnel" where the top methods accept wide types and convert them to known concrete types for which the core implementations are defined on. This means that the core parts will not be invalidated since they have concrete type information while things should still accept generic types.

**Pull request merged**